File: java/dsa/data-structures/binary-tree/diameter-of-binary-tree.java
/** Definition for a binary tree node */
// class TreeNode { int val; TreeNode left; TreeNode right; }

class Solution {
    int diameter = 0;
    
    public int diameterOfBinaryTree(TreeNode root) { // @viz: function_start
        height(root); // @viz: height_start
        return diameter; // @viz: function_start
    }
    
    private int height(TreeNode node) { // @viz: height_start
        if (node == null) return 0; // @viz: base_case
        
        int left = height(node.left); // @viz: recurse_left
        int right = height(node.right); // @viz: recurse_right
        
        diameter = Math.max(diameter, left + right); // @viz: update_diameter
        return 1 + Math.max(left, right); // @viz: return_height
    }
}

File: go/dsa/data-structures/binary-tree/diameter-of-binary-tree.go
/** Definition for a binary tree node */
// type TreeNode struct { Val int; Left, Right *TreeNode }

func diameterOfBinaryTree(root *TreeNode) int { // @viz: function_start
    diameter := 0
    var height func(*TreeNode) int
    height = func(node *TreeNode) int { // @viz: height_start
        if node == nil { return 0 } // @viz: base_case
        
        left := height(node.Left) // @viz: recurse_left
        right := height(node.Right) // @viz: recurse_right
        
        if left + right > diameter { diameter = left + right } // @viz: update_diameter
        if left > right { return 1 + left } // @viz: return_height
        return 1 + right // @viz: return_height
    }
    height(root) // @viz: height_start
    return diameter // @viz: function_start
}